### ISSUE TITLE
Handling NaNs

### ISSUE THREAD

**Issue by vivekthampy** [scikit-beam/scikit-beam#431]
The multi_tau_auto_corr functions returns nothing if there are any NaN values in the array defined by the labelled array. It will be great if we don't have to worry about selecting regions which do not have any NaN values. @sameera2004 @tacaswell @stuwilkins

**Comment by tacaswell** [comment]
This is a feature to allow marking bad frames with nan.

If you have fixed bad pixels just make sure they are removed from your roi
mask.

**Comment by ericdill** [comment]
For an example of the use case that @tacaswell is talking about: https://github.com/scikit-beam/scikit-beam-examples/blob/master/demos/time_correlation/Removing_bad_images_one_time_correlation.ipynb

**Comment by stuwilkins** [comment]
@ericdill that wasn't Viveks comment, he has pixels in the ROI which are always bad and marked as NaNs.

@vivekthampy you could fix this by multiplying your ROI by the flat field checking for NaN. For example

```python
np.ones((960,960), dtype=np.int64) * (np.isnan(np.rot90(new_flat, -1)) == False)
```

Would produce a 960 by 960 mask of 1 and 0 where the pixels are bad. Multiplying this by your ROI should solve the problem. Let me know if it works. Perhaps we should make a useful function for that?

**Comment by stuwilkins** [comment]
P.S. Where ```new_flat``` is the flat field correction.

**Comment by ericdill** [comment]
> @ericdill that wasn't Viveks comment, he has pixels in the ROI which are always bad and marked as NaNs.

@stuwilkins I was merely trying to provide a little more context to the response of @tacaswell

**Comment by sameera2004** [comment]
Yes multi tau auto correlation code assume that you don't have bad pixels, you mask out bad pixels. @tacaswell and @ericdill are talking about when you have bad frames you can convert them to NAN array and it will be taken care as in the example @ericdill was mentioning.

 If you think it is useful after creating the rois you can check the rois.
We can add that check inside the correlation code if you think it is useful??
 Right now CHX correlation pipeline check
1) the rois after creating it
2 ) check  if any of rois the number of pixels are zero before using multi tau auto correlation.

**Comment by mpmdean** [comment]
Hey @vivekthampy,

Just to add to the explanation (in a hopefully helpful way) NaNs are used in the algorithm to tag and not use bad _frames _ (not bad CCD regions) in a proper way.

Start by either indexing the data array to throw away bad regions or choosing the roi to avoid NaNs.

So if there is data spolit by some issue set those frames to have NaN's and continue to send that data it.

Best,
Mark

**Comment by rayosborn** [comment]
I know nothing of the code so ignore this comment if it's not helpful, but Numpy has good support for masked arrays, so that you can exclude bad pixels without explicitly changing the pixel values. Numpy functions will just ignore the masked values when calculating averages, etc. Matplotlib supports them as well.

**Comment by tacaswell** [comment]
masked arrays can add significant performance issues and do not translate out to c-extensions an easily.   

The way this code works one of the inputs is a label array (effectively a (non-binary) mask).

**Comment by ambarb** [comment]
Can we revist this?  There are portions of the fastccd that have NaN's after the flatfield correction that should be no be considered in the ananlysis, but this is per pixel, not per frame. So this dataset is completely ignored.  In other words, there are two flavors of NaN 
* particular pixels (or frames) that are marked with NaN to exclude specific frames from the analysis
* always "NaN" for some other reason (beamstop, dead pixel, saturated pixels)

So it seems we need two different ways to handle NaN. 

Maybe the change in functionality should be to drop the pixel from the entire analysis if that pixel is NaN in all frames, but the analysis would still proceed as if the pixel was not part of a given roi.  OF course, the other way around this is to have customized user code in a notebook that omits the "always NaN" pixel from the ROI mask, but in view of Xi-Cam, I am not sure if this is something that would work well with that environment.  Plus, I think the proposed functionality is nice for the users just getting started.

**Comment by mpmdean** [comment]
Hi @ambarb This issue did cause me some momentary confusion in the past when I got all NaNs returned. 

Despite this, I would worry a bit about trying to make scikit-beam solve this issue, which is essentially an error that the user/program calling the code is passing the wrong ROI. My experience is that one *must* consider the ROIs used for XPCS very carefully in order to avoid misleading results. i.e. one can easily pass in regions that have average intensity variations that aren't coming from speckle. The current beamstop correction is also not something that I would feel comfortable not examining carefully is often hard to get the half-blocked pixels around the edge of the beamstop excluded properly. 

If the user inputs data with NaNs in all frames, the result will be all NaNs, in which case I think it would be good to raise a more descriptive error and leave it at that.

**Comment by tacaswell** [comment]
I think a good middle ground would be a function that given a mask and a full data set returns an updated mask to exclude always NaN pixels?

**Comment by ambarb** [comment]
@mpmdean 
These are good points.  But with the camera behaving as it is, there are random places in the image of the scattering pattern that are NaN.  It is the same for all images.  Of course, you are 100% right that the ROI selection is really important, but I don't see how NaNs have any bearing on that.  

As @stuwilkins suggested above, it is just one line of code in a notebook for me to move forward so I am wondering why we don't add this function to the library.  People will still have to select ROIs with care. And of course, I would never blindly suggest to anyone to include the beamstop without first testing spill over or iteratively changing the ROI until there is a "good feeling" for how those things contribute to the analysis.  

With the NaN feature added, we get upper threshold for free - which means if you have a static speckle that is very strong, you can get around it easily to analyze the other speckles that are much less intense.  The contrast will be 100% incorrect, but it could be useful to know if you need to select better multi-tau parameters that do not hide any decorrelation with a low contrast.  Personally, I am very uncomfortable with changing pixel values to an "average" value  as a work around (either for thresholding or masking NaN pixels because of the flat field that are no where near the beamstop or the wire that holds it)

**Comment by ambarb** [comment]
@tacaswell for your suggestion, are you saying to essentially recreate the mask array for the rois to by excluding always NaN pixels from the original ROI so that analysis goes forward without the NaN pixel.

**Comment by ambarb** [comment]
To be clear, I was not suggesting to get rid of the feature to drop frames from the analysis using NaN.  I am asking for an additional feature, but it has to be applied to the mask roi, not deep down in the code so that no one knows it is happening

**Comment by tacaswell** [comment]
Ah, I now understand the problem, sorry for being dense.  We are looking at the _whole_ image to decide if we should skip the frame, not if any pixel in an ROI is nan :man_facepalming: .

**Comment by ambarb** [comment]
https://github.com/scikit-beam/scikit-beam/issues/563 has related discussion
